### PR TITLE
Adding testcases for oidc_provider

### PR DESCRIPTION
**Pull request by vivekhiwarkar**
### What does this PR change?

- This PR contains tests for oidc_provider.go

### Does the PR depend on any other PRs or Issues? If yes, please list them.

- No

### Checklist

I confirm, that I have...

- [X] Read and followed the contributing guide in `CONTRIBUTING.md`
- [ ] Added tests for this PR
- [X] Formatted the code using `go fmt` (if applicable)
- [ ] Updated [documentation on the Paralus docs site](https://github.com/paralus/website/blob/main/docs) (if applicable)
- [ ] Updated `CHANGELOG.md`

### CODE REVIEW

File: pkg/service/oidc_provider_test.go [hunk 1]
```diff
@@ -0,0 +1,186 @@
...
+	if provider.GetMetadata().GetName() != "user-"+uuuid {
+		t.Error("invalid name returned")
+	}
+	t.Log("PROVIDER: ", provider.GetSpec().GetProviderName())
+	if provider.GetSpec().GetProviderName() != "provider-"+pruuid {
+		t.Error("invalid provider name returned")
+	}
+}
+
+func TestOidcCreateProvider(t *testing.T) {
```
Comment (niravparikh05) [inline pkg/service/oidc_provider_test.go:24]: can you add one more test case for oidc provider create with Duplicate issuer url

File: pkg/service/oidc_provider_test.go [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+	}
+
+	_, err := OP.GetByName(context.Background(), provider)
+	if err != nil {
+		t.Fatal("could not get partner:", err)
+	}
+	performOidcProviderBasicChecks(t, provider, uuuuid, pruuid)
+}
+
+func TestOidcProviderUpdate(t *testing.T) {
```
Comment (niravparikh05) [inline pkg/service/oidc_provider_test.go:108]: Can you add one more test case with invalid issuer / mapper url to cover validateURL function as well.

File: pkg/service/oidc_provider_test.go [hunk 3]
```diff
@@ -0,0 +1,186 @@
...
+		WillReturnResult(sqlmock.NewResult(1, 1))
+
+	provider := &systemv3.OIDCProvider{
+		Metadata: &v3.Metadata{Id: uuuid, Name: "user-" + uuuid},
+		Spec:     &systemv3.OIDCProviderSpec{ProviderName: "provider-" + pruuid},
+	}
+
+	err := OP.Delete(context.Background(), provider)
+	if err != nil {
+		t.Fatal("could not delete partner:", err)
```
Comment (niravparikh05) [inline pkg/service/oidc_provider_test.go:161]: please correct this message. Should be oidc provider instead of partner

File: pkg/service/oidc_provider_test.go [hunk 4]
```diff
@@ -0,0 +1,186 @@
...
+
+	mock.ExpectQuery(`SELECT "oidcprovider"."id", "oidcprovider"."name"`).
+		WithArgs().WillReturnRows(sqlmock.NewRows([]string{"id"}).AddRow(uuuid))
+
+	provider := &systemv3.OIDCProviderList{}
+
+	provider, err := OP.List(context.Background())
+
+	if err != nil {
+		t.Fatal("could not delete partner:", err, pruuid, uuuid)
```
Comment (niravparikh05) [inline pkg/service/oidc_provider_test.go:183]: please correct this message. This is part of testing List

File: pkg/service/oidc_provider_test.go [hunk 5]
```diff
@@ -0,0 +1,186 @@
...
+	db, mock := getDB(t)
+	defer db.Close()
+
+	pruuid := uuid.New().String()
+	uuuid := uuid.New().String()
+
+	OP := NewOIDCProviderService(db, "", getLogger())
+
+	mock.ExpectQuery(`SELECT "oidcprovider"."id", "oidcprovider"."name"`).
+		WithArgs().WillReturnRows(sqlmock.NewRows([]string{"id"}).AddRow(uuuid))
```
Comment (niravparikh05) [inline pkg/service/oidc_provider_test.go:176]: let's add multiple rows here with an assertion below.

File: pkg/service/oidc_provider_test.go [hunk 6]
```diff
@@ -0,0 +1,186 @@
...
+	pruuid := uuid.New().String()
+
+	scope := []string{"system", "local"}
+
+	puuid, uuuid := addParterOrgFetchExpectation(mock)
+
+	mock.ExpectQuery(`SELECT "oidcprovider"."id", "oidcprovider"."name", .* FROM "authsrv_oidc_provider" AS "oidcprovider" WHERE`).
+		WithArgs().WillReturnRows(sqlmock.NewRows([]string{"id"}).AddRow(uuuid))
+
+	mock.ExpectExec(`UPDATE "authsrv_oidc_provider"`).
```
Comment (niravparikh05) [inline pkg/service/oidc_provider_test.go:124]: expand this mock query for just the given oidc provider id, refer [here](https://github.com/paralus/paralus/blob/main/pkg/service/role_test.go#L219)

**Review by niravparikh05 — DISMISSED** [review]

File: pkg/service/oidc_provider_test.go [hunk 7]
```diff
@@ -0,0 +1,254 @@
...
+
+	mock.ExpectQuery(`SELECT "oidcprovider"."id", "oidcprovider"."name", .* FROM "authsrv_oidc_provider" AS "oidcprovider" WHERE`).
+		WithArgs().WillReturnRows(sqlmock.NewRows([]string{"id"}).AddRow(uuuuid))
+
+	mock.ExpectExec(`UPDATE "authsrv_oidc_provider" AS "oidcprovider" SET "name" = 'user-` + uuuid + `', .*"organization_id" = '` + ouuid + `', "partner_id" = '` + puuid + `.* WHERE \(id = '` + uuuuid + `'\)`).
+		WillReturnResult(sqlmock.NewResult(1, 1))
+
+	provider := &systemv3.OIDCProvider{
+		Metadata: &v3.Metadata{Id: uuuuid, Name: "user-" + uuuid, Partner: "partner-" + puuid, Organization: "org-" + ouuid},
+		Spec:     &systemv3.OIDCProviderSpec{Scopes: scope, IssuerUrl: "test.url", ProviderName: "provider-" + pruuid, MapperUrl: "test.url"},
```
Comment (meain) [inline pkg/service/oidc_provider_test.go:191]: I think we could probably create a [table driven test](https://mj-go.in/golang/table-driven-tests-in-go) for the error cases to test them individually.

File: pkg/service/oidc_provider_test.go (hunk 2 above)
Comment (meain) [reply pkg/service/oidc_provider_test.go:137]: Started with in `TestOidcProviderUpdateInvalidUrl`

File: pkg/service/oidc_provider_test.go [hunk 8]
```diff
@@ -0,0 +1,254 @@
...
+	pruuid1 := uuid.New().String()
+	pruuid2 := uuid.New().String()
+
+	OP := NewOIDCProviderService(db, "", getLogger())
+
+	mock.ExpectQuery(`SELECT "oidcprovider"."id", "oidcprovider"."name".* FROM "authsrv_oidc_provider" AS "oidcprovider"`).
+		WithArgs().WillReturnRows(sqlmock.NewRows([]string{"id", "name"}).
+		AddRow(pruuid, "provider_name-"+pruuid).
+		AddRow(pruuid1, "provider_name-"+pruuid1).
+		AddRow(pruuid2, "provider_name-"+pruuid2))
```
Comment (meain) [inline pkg/service/oidc_provider_test.go:241]: Maybe add one more field, say `issuerUrl`.

File: pkg/service/oidc_provider_test.go [hunk 9]
```diff
@@ -0,0 +1,254 @@
...
+
+	OP := NewOIDCProviderService(db, "", getLogger())
+
+	pruuid := uuid.New().String()
+	uuuid := uuid.New().String()
+
+	mock.ExpectQuery(`SELECT "oidcprovider"."id", "oidcprovider"."name"`).
+		WithArgs().WillReturnRows(sqlmock.NewRows([]string{"id"}).AddRow(uuuid))
+
+	mock.ExpectExec(`UPDATE "authsrv_oidc_provider"`).
```
Comment (meain) [inline pkg/service/oidc_provider_test.go:212]: Probably add more of the query here than just an `UPDATE` call.

**Review by meain — COMMENTED** [review]
Few more changes.

There are some pending items from previous review. I have marked the stale comments as resolved. Make sure to do this exercise the next time you have updated the PR so that it is easier to track the changes made, plus you will be able to make sure you have hit all the requested changes.

File: pkg/service/oidc_provider_test.go [hunk 10]
```diff
@@ -0,0 +1,283 @@
...
+		WithArgs().WillReturnError(fmt.Errorf("unique constraint violation"))
+
+	provider := &systemv3.OIDCProvider{
+		Metadata: &v3.Metadata{Partner: "partner-" + puuid, Organization: "org-" + ouuid, Name: "user-" + uuuid},
+		Spec:     &systemv3.OIDCProviderSpec{Scopes: scope, IssuerUrl: "https://token.actions.githubusercontent.com", ProviderName: "provider-" + pruuid},
+	}
+
+	provider, err := OP.Create(context.Background(), provider)
+	if err == nil {
+		t.Fatal("expected create provider fail, but was created")
```
Comment (meain) [inline pkg/service/oidc_provider_test.go:51]:
```suggestion
		t.Fatal("expected create provider fail on duplicate issuer url, but was created")
```

File: pkg/service/oidc_provider_test.go [hunk 11]
```diff
@@ -0,0 +1,283 @@
...
+	v3 "github.com/paralus/paralus/proto/types/commonpb/v3"
+	systemv3 "github.com/paralus/paralus/proto/types/systempb/v3"
+)
+
+func performOidcProviderBasicChecks(t *testing.T, provider *systemv3.OIDCProvider, uuuid string, pruuid string) {
+	if provider.GetMetadata().GetName() != "user-"+uuuid {
+		t.Error("invalid name returned")
+	}
+	if provider.GetSpec().GetProviderName() != "provider-"+pruuid {
+		t.Error("invalid provider name returned")
```
Comment (meain) [inline pkg/service/oidc_provider_test.go:19]: It would help to check mapper_url , issuer_url, etc here.

Comment (meain) [reply pkg/service/oidc_provider_test.go:19]: Not sure why this was marked as resolved. The idea here to check for all the values the the `oidc_provider` returns.

File: pkg/service/oidc_provider_test.go [hunk 12]
```diff
@@ -0,0 +1,186 @@
...
+
+	mock.ExpectQuery(`SELECT "oidcprovider"."id" FROM "authsrv_oidc_provider" AS "oidcprovider" WHERE .organization_id = '` + ouuid + `'. AND .partner_id = '` + puuid + `'. AND .name = 'user-` + uuuid + `'.`).
+		WillReturnError(fmt.Errorf("no data available"))
+
+	scope := []string{"system"}
+
+	mock.ExpectQuery(`SELECT "oidcprovider"."id"`).
+		WillReturnError(fmt.Errorf("no data available here"))
+
+	mock.ExpectQuery(`INSERT INTO "authsrv_oidc_provider"`).
```
[the remark this answers is not in the document]
Comment (meain) [reply pkg/service/oidc_provider_test.go:41]: Not sure why this was marked as resolved. The idea is to update the `INSERT` query to include more of the query than just the table name.

**Review by meain — APPROVED** [review]
LGTN
